admin/aws.py

```diff
@@ -471,7 +480,45 @@ def _perform_upload_s3_key(self, dispatcher, intent):
         """
         bucket = self.s3_buckets[intent.target_bucket]
         with intent.file.open() as source_file:
-            bucket[intent.target_key] = source_file.read()
+            bucket[intent.target_key] = source_file.read(), intent.content_type
+
+    def get_bucket_keys(self, bucket):
+        """
+        Retrieve the keys from a bucket in S3.
+
+        :param str bucket: name of the S3 bucket.
+        :returns: the keys present in the bucket.
+        """
+        return self.s3_buckets[bucket].keys()
+
+    def get_bucket_contents(self, bucket):
+        """
+        Retrieve the file keys and contents in S3.
+
+        :param str bucket: name of the S3 bucket.
+        :returns: a dictionary mapping keys to file contents.
+        """
+        return {k: v[0] for k, v in self.s3_buckets[bucket].items()}
+
+    def get_content(self, bucket, key):
+        """
+        Retrieve the content of a file in S3.
+
+        :param str bucket: name of the S3 bucket.
+        :param str key: key for the S3 file.
+        :returns: the contents of the file.
+        """
+        return self.s3_buckets[bucket][key][0]
+
+    def get_content_type(self, bucket, key):
+        """
+        Retrieve the content of a file in S3.
+
+        :param str bucket: name of the S3 bucket.
+        :param str key: key for the S3 file.
+        :returns: the contents of the file.
+        """
+        return self.s3_buckets[bucket][key][1]
```

Comment:
Removed all these changes in favour of a subclass of `unicode` that has an additional `content_type` attribute. This allows all existing tests to work as before.